### DISCUSSION TITLE
Support for Specifying Schema for Individual Tables

### DISCUSSION THREAD

**Discussion by mverde-taal** [prisma/prisma#24920] [Q&A] [answered]
I would like to ask if it's possible to implement a feature that allows specifying the schema only for certain tables within the schema.prisma file, while using the default schema defined in the connection string for other tables.

Context
I am developing a multi-tenant application where each tenant has its own schema, but there are also some common tables shared across all tenants, which reside in a separate common schema.

Currently, it is possible to specify different schemas for different tables, but they must be specified beforehand. This approach requires defining all tables within the same schema.prisma file to belong to the schema defined in the connection string, or defining them in separate schema.prisma files with separate connections.

With this new feature, I could specify the tenant's schema in the connection string while always having access to the common tables. This would resolve the issue of having to create two separate PrismaService instances, which currently prevents me from performing joins between tenant-specific tables and common tables.

Proposal
I propose the implementation of a feature that allows specifying the schema at the table level in the schema.prisma file, as illustrated below:

```
datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

generator client {
  provider = "prisma-client-js"
}

model User {
  id        Int     @id @default(autoincrement())
  name      String
  email     String  @unique
  lookupId  Int
  lookupTable LookupTable @relation(fields: [lookupId], references: [id])
}

model LookupTable {
  id    Int     @id @default(autoincrement())
  name  String
  value String
  @@map("common.lookup_table") // Schema specification
}
```
In this way, if the connection string points to a specific tenant schema (e.g., tenant1), all tables without a specified schema will use that schema. However, the LookupTable will reside in the common schema.

Benefits
Increased Flexibility: Allows defining tables residing in different schemas within the same schema.prisma file.
Better Management of Common Data: Facilitates managing common tables shared across different tenant schemas.
Fewer Separate Configurations: Reduces the need for multiple schema.prisma files and multiple Prisma clients, simplifying configuration and maintenance.
Enable Joins Across Schemas: Allows performing joins between tenant-specific tables and common tables without needing to manage multiple PrismaService instances.
Considerations
This change should be backward compatible, as tables without a specified schema will continue to use the schema defined in the connection string.
It might be useful to also have a way to specify a default schema at the schema.prisma file level.
Thank you for your attention and for the exceptional work you are doing with Prisma. I hope this feature can be considered for a future release.

Best regards,

**Comment by ludralph** [comment]
Hi @mverde-taal 👋 

It is indeed possible to specify different schemas for different tables within the same `schema.prisma` file. This can be achieved using the [`@@schema`](https://www.prisma.io/docs/orm/prisma-schema/data-model/multi-schema) attribute for each model that belongs to a specific schema.

Here is an example of how you can define tables residing in different schemas within the same schema.prisma file:
```prisma
datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
  schemas  = ["common", "tenant"]
}

generator client {
  provider = "prisma-client-js"
  previewFeatures = ["multiSchema"]
}

model User {
  id        Int     @id @default(autoincrement())
  name      String
  email     String  @unique
  lookupId  Int
  lookupTable LookupTable @relation(fields: [lookupId], references: [id])
  @@schema("tenant") 
}

model LookupTable {
  id    Int     @id @default(autoincrement())
  name  String
  value String
  @@schema("common") 
}
```

**Reply by mverde-taal** [reply]
Thank you for your response.

However, my use case involves dynamically creating and managing schemas for tenants that are not predefined and whose number is not known in advance. We handle this using a middleware to modify the schema in the connection string based on the tenant ID extracted from the request. Here is a simplified version of how we achieve this:

```
import { Injectable, NestMiddleware } from '@nestjs/common';
import { Request, Response, NextFunction } from 'express';

@Injectable()
export class TenantMiddleware implements NestMiddleware {
  use(req: Request, res: Response, next: NextFunction) {
    const tenantId = req.headers['x-tenant-id'] as string;
    if (!tenantId) {
      return res.status(400).send('Tenant ID is required');
    }

    // Attach the tenantId to the request object
    req['tenantId'] = tenantId;
    next();
  }
}

```
In this way, the tenant's schema is dynamically set in the connection string. However, this approach makes it difficult to access a common schema that contains shared tables.
The current limitation we face is that we cannot dynamically specify the schema for tenant-specific tables while also having constant access to the common tables.
We have to create two services PrismaCommonService and PrismaTenantService, but this prevent to join tables of different services.

We need a feature that allows specifying the schema for some tables within the schema.prisma file while using the schema from the connection string for other tables. This way, we can dynamically set the tenant schema in the connection string and still access common tables.

Here is an example of what we are looking to achieve:

```
datasource db {
  provider = "postgresql://user:password@servername:5432/dbname?schema=`{tenantId}`"
  url      = "p
}

generator client {
  provider = "prisma-client-js"
}

//this use the schema in the connection string
model TenantTable {
  id        Int     @id @default(autoincrement())
  name      String
  email     String  @unique
  lookupId  Int
  lookupTable LookupTable @relation(fields: [lookupId], references: [id])
}

// This should always use the 'common' schema
model LookupTable {
  id    Int     @id @default(autoincrement())
  name  String
  value String
  @@schema("common") 
}

```
Benefits
Increased Flexibility: Allows defining tables that reside in different schemas within the same schema.prisma file, making it easier to manage multi-tenant setups with shared common tables.
Dynamic Schema Handling: Facilitates dynamic tenant schema management while providing constant access to common tables.
Simplified Configuration: Reduces the need for multiple schema.prisma files and multiple Prisma clients, simplifying the configuration and maintenance process.

**Reply by ludralph** [reply, marked as answer]
Hi @mverde-taal, Can you please create a [feature request](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=feature_request.md&title=) and document your use case so that our Engineers can take a look?
